enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,182 @@
+
+
+### Implementation Details/Notes/Constraints
+
+The `external CCM` implementation will be hosted within openshift repository in https://github.com/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers + `Openstack` out-of-tree provider implementation.
+
+`Cloud-controller` will be provisioned in the `openshift-cloud-controller` namespace.
+
+Preparation steps, common for all providers:
+- Add support for recognition of the `external` option in library-go: https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L154 which will disable config parsing and add `external-cloud-config` option in the `Infrastructure` resource, which will be only used for external providers.
+- Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to newly created `openshift-cloud-controller` namespace and be managed with an operator.
+
+Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:
+- Add `external` flag to `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.
+- Add `external` flag to `kubelet`. Remove the `cloud-config` option from the template.
+- Deploy a `cloud-controller` pod in the `openshift-cloud-controller` namespace.
+
+## Design Details
+
+### Kubelet
+
+Openshift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are passed externally from the `Infrastructure` resource on https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357, where the support for `external` cloud provider should be added. For each platform with `external` configuration, the `cloud-config` flag should not be set.
+
+### Kube-api-server
+
+No changes for the current configuration are needed
+
+### Kube-controller-manager
+
+TBD
+
+#### Upgrade/Downgrade strategy
+
+Upstream [proposal](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#motivation) describes main pinpoints for migration between in-tree and out-of-tree for HA clusters. They propose implementation for cross-namespace leadership delegation, which would help Openshift to preserve HA during upgrages.
+
+In case the cluster does not require to be HA, then upgrading under CVO management, which will lead to two simultanious replicas of `cloud-controllers` running at the same time at some moment is not a problem. Each provider, therefore supports multiple options to run the `cloud-controller`:
```

Comment:
> running at the same time at some moment is not a problem.

This part of this statement should be justified.

> When two replicas of the `cloud-controllers` are running simultaneously, <something happens and that's ok because...>